trl/trainer/ppo_trainer.py

```diff
@@ -265,17 +265,43 @@ def __init__(
         else:
             self.kl_ctl = FixedKLController(self.config.init_kl_coef)
 
-        (
-            self.model,
-            self.ref_model,
-            self.optimizer,
-            self.data_collator,
-            self.dataloader,
-            self.lr_scheduler,
-        ) = self.accelerator.prepare(
-            self.model, self.ref_model, self.optimizer, self.data_collator, self.dataloader, self.lr_scheduler
+        # Safety checkers for DS integration
+        is_deepspeed_used = self.accelerator.distributed_type == "DEEPSPEED" and hasattr(
+            self.accelerator.state, "deepspeed_plugin"
         )
 
+        if is_deepspeed_used:
+            (
+                self.model,
+                self.optimizer,
+                self.data_collator,
+                self.dataloader,
+                self.lr_scheduler,
+            ) = self.accelerator.prepare(
+                self.model, self.optimizer, self.data_collator, self.dataloader, self.lr_scheduler
+            )
+            # 8 bit models are already set on the correct device
+            if not getattr(self.ref_model.pretrained_model, "is_loaded_in_8bit", False):
+                # DS integration only allows for single model and as `ref_model` is only used for
+                # `KL devergence loss`,i.e, in eval model, just have it be on the respective device and
+                # there is no need to pass it to the `accelerator.prepare` call
+                self.ref_model = self.ref_model.to(self.accelerator.device)
+
+            # this hack seems to be needed for DS stage 3 to work
+            if self.accelerator.state.deepspeed_plugin.zero_stage == 3:
+                self.model.train()
```

Comment:
Based on the offline discussion I had with @pacman100 , I confirm this hack is needed.